### PR TITLE
Refactor public API analyzer, banned API analyzer and version check a…

### PR DESCRIPTION
**Pull request by mavasani**
…nalyzer into separate packages under DotNetAnalyzers.

We decided in a design meeting that we will retain these analyzers in this repo. New analyzer packages:
1. DotNetAnalyzers.PublicApiAnalyzer: Contains separate assemblies for analyzers and code fixes
2. DotNetAnalyzers.BannedApiAnalyzer
3. DotNetAnalyzers.VersionCheckAnalyzer (package referenced by FxCop analyzers package to generate a diagnostic if version of Microsoft.CodeAnalysis loaded at runtime is less than the version of Microsoft.CodeAnalysis referenced at compile time)

Roslyn.Diagnostics.Analyzers references DotNetAnalyzers.PublicApiAnalyzer and DotNetAnalyzers.BannedApiAnalyzer.
Microsoft.CodeAnalysis.FxCopAnalyzers references DotNetAnalyzers.VersionCheckAnalyzer

TODOs for follow-up PRs:
1. Move DotNetAnalyzers.PublicApiAnalyzer and DotNetAnalyzers.VersionCheckAnalyzer to reference 1.x version of Microsoft.CodeAnalysis
2. Switch DotNetAnalyzers.PublicApiAnalyzer analyzer assembly to reference Microsoft.CodeAnalysis.CSharp NuGet package instead of Microsoft.CodeAnalysis NuGet package.

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @AArnott

**Comment by mavasani** [comment]
@sharwell @AArnott - let me know if you have strong concerns on the current package structure here.

**Comment by Pilchie** [comment]
Agree with @sharwell on naming. Should the IDs change to not be `RS`?

**Comment by mavasani** [comment]
Going to update the naming with the next commit.

> Should the IDs change to not be `RS`

I thought about it, but noticed that even Sam's fork of public API analyzer shares same RS IDs. I also noted quite a bunch of suppressions and ruleset configuration for RS rules in Roslyn.sln, and I presume same holds for all client's of Sam's forked package. If we are okay to break these suppressions/configurations, then I can do the change. Not sure if the downstream breaks are worth for better choice of IDs though.

**Comment by mavasani** [comment]
@sharwell Addressed feedback.

File: src/Microsoft.CodeAnalysis.VersionCheckAnalyzer/VersionCheckAnalyzer.cs [hunk 1]
```diff
@@ -53,11 +53,11 @@ public override void Initialize(AnalysisContext context)
...
                 compilationStartContext.RegisterCompilationEndAction(compilationContext =>
                 {
-                    if (!s_ShouldExecuteOperationAnalyzers)
+                    if (!s_ShouldExecuteAnalyzers)
                     {
                         // Version mismatch between the analyzer package '{0}' and Microsoft.CodeAnalysis '{1}'. Certain analyzers in this package will not run until the version mismatch is fixed.
-                        var arg1 = AnalyzerPackageName + "-" + AnalyzerPackageVersion;
-                        var arg2 = s_ShouldExecuteOperationAnalyzers;
+                        var arg1 = RequiredMicrosoftCodeAnalysisVersion;
+                        var arg2 = s_ShouldExecuteAnalyzers;
```
Comment (genlu) [inline src/Microsoft.CodeAnalysis.VersionCheckAnalyzer/VersionCheckAnalyzer.cs]: Argument to the message seems wrong
```suggestion
                        var arg2 = s_MicrosoftCodeAnalysisVersion;
```

Comment (mavasani) [reply src/Microsoft.CodeAnalysis.VersionCheckAnalyzer/VersionCheckAnalyzer.cs]: Thanks, sounds good. This analyzer still doesn't work. Need to port it to reference Microsoft.CodeAnalysis 1.0 to work.

**Comment by mavasani** [comment]
Thanks @sharwell @genlu

**Pull request merged**